### PR TITLE
hashicorp acceptance requirements

### PR DESCRIPTION
**Pull request by arykalin**
@tyrannosaurus-becks could you clarify this requirement:
```
Consider writing one test that exercises every API endpoint of backend like this. Then consider combining those tests into a full run of the backend like this. The idea here is to try every endpoint in the order one would expect them in real life, and verify from the outside that expected responses are being returned.
```

Do you mean that we should use something like this:
```
	req := &logical.Request{
		Operation: logical.ReadOperation,
		Path:      "config",
		Storage:   e.Storage,
	}
```
instead of cluster := vault.NewTestCluster we have currently in https://github.com/Venafi/vault-pki-backend-venafi/blob/master/plugin/pki/backend_test.go?

### CODE REVIEW

**Comment by tyrannosaurus-becks** [comment]
Hi! Ah, cool to hear you're looking at that. 

I was meaning something like [this particular test](https://github.com/hashicorp/vault-plugin-secrets-alicloud/blob/master/backend_test.go#L134). It sets up the back-end, and then each test below is one call. Every call type the back-end could receive is exercised, and each field is read.

It basically ensures that each endpoint is operating the way we'd expect, and that when configured correctly, secrets are issued, and can be renewed and revoked.

**Comment by arykalin** [comment]
@tyrannosaurus-becks could you clarify, does it mean that it is preferable to use `logical.Backend` and `Backend.HandleRequest`  for testing instead of `vault.NewTestCluster` and `cluster.Cores[0].Client`? Or you talking only about test structure, where we firstly setup `logical.Backend` or `vault.NewTestCluster` and then run tests on them?

**Comment by tyrannosaurus-becks** [comment]
@arykalin 

>does it mean that it is preferable to use logical.Backend and Backend.HandleRequest for testing

Yes, it's preferable because it's more of a "black box" test that way. Then you're merely making calls from the "outside" and viewing the responses. It's easy to tell what those calls are because there is only one call and response per test. 

This not only tests that the expected calls get the expected responses, but it has the added benefit of clearly documenting how the plugin is intended to be used and verifying that when used that way, it does work. Then if we end up with a user who is using it totally differently, we can look at the tests and try to guide them back to the way it's intended to be used, or add tests that mimic what they're doing to duplicate their results and do further work.

>instead of vault.NewTestCluster and cluster.Cores[0].Client

We have at times internally discussed no longer using `vault.NewTestCluster` because it can be a bit "magical". We do still use it but only when there's no other apparent alternative. Someday we may heavily refactor it. It would be great to avoid it here.

**Comment by arykalin** [comment]
Thank you @tyrannosaurus-becks 
Could you look into tests   [file](https://github.com/Venafi/vault-pki-backend-venafi/blob/hashi-acceptance-VEN-55982/plugin/pki/backend_test.go) now, are they fetch Hashicorp requirements?
I splitted tests into 3 parts, for fake endpoint (vcert test CA) I'm testing all pki-backend endpoints. For Venafi Platform and Cloud I'm testing only issue and sign to check that integration with Venafi products is working fine.

**Comment by mr-tron** [comment]
tests passed

File: plugin/pki/backend_test.go [hunk 1]
```diff
@@ -1,596 +1,53 @@
...
-	data.cert = resp.Data["certificate"].(string)
-
-	checkStandartCert(t, data)
+	t.Run("fake create role", integrationTestEnv.FakeCreateRole)
+	t.Run("fake list roles", integrationTestEnv.FakeListRole)
+	t.Run("fake issue", integrationTestEnv.FakeIssueCertificate)
+	t.Run("fake sign", integrationTestEnv.FakeSignCertificate)
+	t.Run("fake list certificates", integrationTestEnv.FakeListCertificate)
+	t.Run("fake read certificate by cn", integrationTestEnv.FakeReadCertificateByCN)
+	t.Run("fake read certificate by serial", integrationTestEnv.FakeReadCertificateBySerial)
```
Comment (tyrannosaurus-becks) [inline plugin/pki/backend_test.go]: For the tests where we're creating something, like `FakeCreateRole`, it would be nice to have a test like `FakeReadRole` right afterwards that checks that all the values that were written are returned, and that no sensitive fields like `tpp_password` are returned.

For the tests where we're reading something, like `FakeReadCertificateByCN` and `FakeReadCertificateBySerial`, it would be nice to add a check for each field that's configurable by the user to make sure it's having the effect the user expects.

File: plugin/pki/backend_test.go [hunk 2]
```diff
@@ -1,596 +1,53 @@
...
+	t.Run("fake revoke certificate", integrationTestEnv.FakeRevokeCertificate)
```
Comment (tyrannosaurus-becks) [inline plugin/pki/backend_test.go]: I'm curious about why revocation isn't implemented yet? Should the endpoint still be included if it's not implemented?

Comment (arykalin) [reply plugin/pki/backend_test.go]: We're planning to add revocation in future but for now didn't have time to implement it. You think we should remove it?

File: plugin/pki/backend_test.go (hunk 1 above)
Comment (arykalin) [reply plugin/pki/backend_test.go]: Ok, I added those tests:
	t.Run("tpp create role", integrationTestEnv.TPPCreateRole)
	t.Run("tpp read role", integrationTestEnv.TPPReadRole)
	t.Run("cloud create role", integrationTestEnv.CloudCreateRole)
	t.Run("cloud read role", integrationTestEnv.CloudReadRole)
	t.Run("fake create role", integrationTestEnv.FakeCreateRole)
	t.Run("fake create role", integrationTestEnv.FakeCreateMixedRole)
@tyrannosaurus-becks  could you review?

File: plugin/pki/backend_test.go (hunk 2 above)
Comment (arykalin) [reply plugin/pki/backend_test.go]: @tyrannosaurus-becks we think the "not implemented" error is appropriate since it is a feature that the Vault native PKI supports and we don't have parity.

Comment (tr1ck3r) [reply plugin/pki/backend_test.go]: @tyrannosaurus-becks we didn't implement revocation simply because it isn't a priority for the types of use cases we targeted (i.e. DevOps), same opinions that are expressed around revocation in the [Vault docs](https://www.vaultproject.io/docs/secrets/pki/index.html).  I anticipate that we will add it in the future and, as @arykalin said, feel it is appropriate for "not implemented" error to be returned since POST /pki/revoke is native Vault functionality.

Comment (tyrannosaurus-becks) [reply plugin/pki/backend_test.go]: I see what you're saying. I'm not aware of it being a paradigm followed in Vault. I don't see any harm in keeping it though.
